common/workunit/workflow.cpp

```diff
@@ -613,12 +776,816 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+        }
+    }
+#ifdef TRACE_WORKFLOW
+    //verifying the predecessor counts
+    unsigned totalPredecessors = 0;
+    unsigned totalSuccessors = 0;
+    unsigned totalConditionItems = 0;
+    //iterate through the workflow items
+    for(int i = 1; i <= workflow->count(); i++)
+    {
+        CCloneWorkflowItem & cur = static_cast<CCloneWorkflowItem&>(workflow->queryWfid(i));
+        unsigned numPred = cur.queryNumPredecessors();
+        unsigned numSuc = cur.queryNumSuccessors();
+        totalPredecessors += numPred;
+        totalSuccessors += numSuc;
+        LOG(MCworkflow, "Item %u has %u predecessors and %u successors", cur.queryWfid(), numPred, numSuc);
+
+        if(cur.queryMode() == WFModeCondition)
+        {
+            totalConditionItems++;
+        }
+    }
+    //iterate throught the IntermediaryWorkflow items
+    for(int i = 0;  i < logicalWorkflow.size() ; i++)
+    {
+
+        IRuntimeWorkflowItem  *tmp = logicalWorkflow[i].get();
+        CCloneWorkflowItem * cur = static_cast<CCloneWorkflowItem*>(tmp);
+        unsigned numPred = cur->queryNumPredecessors();
+        unsigned numSuc = cur->queryNumSuccessors();
+        totalPredecessors += numPred;
+        totalSuccessors += numSuc;
+        LOG(MCworkflow, "Runtime item %u has %u predecessors and %u successors", cur->queryWfid(), numPred, numSuc);
+        if(cur->queryMode() == WFModeCondition)
+        {
+            totalConditionItems++;
+        }
+    }
+    LOG(MCworkflow, "Total predecessors is: %u,  total successors is: %u, total condition items is: %u", totalPredecessors, totalSuccessors, totalConditionItems);
+#endif
+}
+CCloneWorkflowItem * WorkflowMachine::insertLogicalPredecessor(unsigned successorWfid)
+{
+    unsigned wfid = workflow->count() + logicalWorkflow.size()+1;
+    #ifdef TRACE_WORKFLOW
+        LOG(MCworkflow, "new predecessor workflow item %u has been created", wfid);
+    #endif
+
+    CCloneWorkflowItem * predecessor = new CCloneWorkflowItem(wfid); //initialise the intermediary
+    Owned<IRuntimeWorkflowItem> tmp = predecessor;
+    logicalWorkflow.push_back(tmp); //adding it to the workflow array
+
+    markDependencies(successorWfid, predecessor, false, false);
+    return predecessor;
+}
+//predecessor is for Condition/Sequential/Ordered
+void WorkflowMachine::markDependencies(unsigned int wfid, CCloneWorkflowItem *logicalPredecessor, bool prevOrdered, bool activate)
+{
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Called mark dependents on item %u", wfid);
+    #endif
+
+    CCloneWorkflowItem & item = static_cast<CCloneWorkflowItem&>(workflow->queryWfid(wfid));
+    if(activate)
+    {
+        item.activate();
+    }
+    //This checks that the current item has not already performed the markDependencies algorithm
+    if(!item.isDependentSuccessorsEmpty())
+    {
+        if(activate)
+        {
+            activateDependencies(item);
+        }
+        return;
+    }
+    if(logicalPredecessor && (!prevOrdered))
```

Comment:
Would you be able to explain why it might not be the correct condition?

Reply:
I think your algorithm has changed since the beginning.  I think it is now
i) an item can be executed if it is active and has no pending dependentPredecessors
ii) When an item completes 
a) all its dependents have their predecessor count decremented, and if it becomes zero then they are run.  (It is an error if they are not active)
b) all of its logical sucessors are marked as active, and added to the queue if they have no dependencies

If so I think the description at the start of the c++ file should be updated to make it really clear.  I'll then look at this again if that assertion is correct.